### PR TITLE
Docs: Contributing eased for beginners in Go, Git and GitHub

### PR DESCRIPTION
**Pull request by olefrost**
#### What is the purpose of this change?

To make it easier to contribute without prior knowledge of Git, GitHub and Go.

The change includes improved/added steps to:
 * Install Git with basic setup
 * Use both SSH and HTTPS for the git origin
 * Install Go and verify the GOPATH
 * Update the forked master
 * Find a popular editor for Go

#### Was the change discussed in an issue or in the forum before?

The addition of an HTTPS link for the git origin was discussed and approved in this forum thread:
https://forum.rclone.org/t/contributing-guide-git-github-com-https-github-com/23855

The rest is based on my own experiences while preparing [my first contribution](https://github.com/rclone/rclone/pull/5273) on a Windows computer. 

I have done my best to research the changes and make the guide applicable for all platforms, protocols and preferences.

I am however still a beginner, so a careful review by somebody with native skills in Go, Git, GitHub and non-Windows platforms is needed. 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have tested the added/modified links.
- [x] I have tested the modified parts of the installation flow on Windows.
- [x] I have tested the document layout in Preview mode on GitHub and Visual Studio Code.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by olefrost** [comment]
There is a minor issue in the proposed push to the forked master. I am not getting the new labels pushed, e.g. v.1.55.1.
I am investigating...

**Comment by olefrost** [comment]
> There is a minor issue in the proposed push to the forked master. I am not getting the new tags pushed, e.g. v1.55.1.

TLDR: There is no issue, the proposed steps are fine.

I am quite sure I have found the explanation: I pushed a test branch without --follow-tags before pushing the master with --follow-tags. The test branch relied on the commit tagged v1.55.1. My later push of the master branch therefore didn’t push the commit tagged v1.55.1 and therefore didn’t transfer the tag.

Lesson learned: If you care about the version tags on the forked master, always push the updated master branch before the development branch, or use –follow-tags whenever pushing or `git config --global push.followTags true`.

Conclusion: The proposed updates are fine; the updated master is pushed before my-new-feature 😅

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -23,28 +23,49 @@ with the [latest beta of rclone](https://beta.rclone.org/):
...
     git remote rename origin upstream
+      # if you have SSH keys setup in your GitHub account:
     git remote add origin git@github.com:YOURUSER/rclone.git
-    go build
+      # otherwise:
+    git remote add origin https://github.com/YOURUSER/rclone.git
+
+Note that most of the terminal commands in the rest of this guide must be executed from the rclone folder created above.
+
+Now [install Go](https://golang.org/doc/install) and verify the [GOPATH](https://github.com/golang/go/wiki/GOPATH) is pointing to a suitable folder:
```
Comment (ncw) [inline CONTRIBUTING.md:45]: I don't think it is necessary to set GOPATH any more to build rclone with modern Go compilers.

Did you have to set it?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -23,28 +23,49 @@ with the [latest beta of rclone](https://beta.rclone.org/):
...
+
+Great, you can now compile and execute your own version of rclone:
+
+    go build                              
+    rclone version
+
+Finally update your forked code to the latest version of the [rclone master](https://github.com/rclone/rclone/tree/master):
+
+    git checkout master
+    git fetch upstream
```
Comment (ncw) [inline CONTRIBUTING.md:57]: I'd normally recommend `git pull --rebase upstream` for Git newbies here instead of

    git fetch upstream
    git merge --ff-only

Especially if you've been working on master.

I guess `git pull --rebase` potentially leaves you sorting out a conflict whereas `git merge --ff-only` will refuse the merge.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -23,28 +23,49 @@ with the [latest beta of rclone](https://beta.rclone.org/):
...
+Now [install Go](https://golang.org/doc/install) and verify the [GOPATH](https://github.com/golang/go/wiki/GOPATH) is pointing to a suitable folder:
+
+    go env GOPATH
+
+Great, you can now compile and execute your own version of rclone:
+
+    go build                              
+    rclone version
+
+Finally update your forked code to the latest version of the [rclone master](https://github.com/rclone/rclone/tree/master):
```
Comment (ncw) [inline CONTRIBUTING.md:54]: This should already be the latest checkout should it since we just `git clone`d the repo.

Perhaps this should be rephrased "To update your forked..."

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -23,28 +23,49 @@ with the [latest beta of rclone](https://beta.rclone.org/):
...
+
+    git checkout master
+    git fetch upstream
+    git merge --ff-only
+    git push origin --follow-tags
```
Comment (ncw) [inline CONTRIBUTING.md:59]: `--follow-tags`? I don't think contributors should need to push tags should they?

**Review by ncw — COMMENTED** [review]
Nice set of updates - thanks!

I put some comments inline for discussion.

File: CONTRIBUTING.md (hunk 1 above)
Comment (olefrost) [reply CONTRIBUTING.md:45]: Good point, there is no need to set I specifically. GOTPATH is automatically set to $HOME/go on Unix and %USERPROFILE%\go on Windows.

How about changing the text to:

File: CONTRIBUTING.md (hunk 2 above)
Comment (olefrost) [reply CONTRIBUTING.md:57]: As a beginner I really like the transparency of first fetching from remote to local repo, then comparing local repo to my workspace and then merging the local repo to my workspace. It helped me tremendously in building my conceptual model of Git.

I was utterly confused when I first read about git pull before understanding the basic principles of fetch, merge and the difference between remote, local and working space. Just too much in one mouthful for me.

I therefore intentionally split the commands to ease the Git understanding for curious contributors googling the commands before execution. I guess experienced Gitters will just do a corresponding pull of their own liking.

You are right, `–ff-only` will refuse the merge if you accidently worked in master. In this situation I prefer to fail-fast (and hard). I haven’t tried, but I assume the mess is smaller and therefore easier to clean-up compared to the situation after a (series of) rebases (with or without conflicts). 

I’m also worried that (thoughtless) rebasing could lead to issues in later work. Suppose a beginner overlooks a misplaced commit in a successfully rebased master. Couldn’t that lead to troubles sometime later?

These were the (possibly naive) thoughts behind my proposal. You are the one with the experience in Git and the typical needs/mistakes of contributors, so I will leave the decision to you.

File: CONTRIBUTING.md (hunk 3 above)
Comment (olefrost) [reply CONTRIBUTING.md:54]: I fully agree, it should be changed. See proposal below.

File: CONTRIBUTING.md (hunk 4 above)
Comment (olefrost) [reply CONTRIBUTING.md:59]: You have a point and I think the real killer question is: Do contributors have any benefit from keeping their cloned master in GitHub up to date? After playing a bit more with GitHub my answer is “No!”, there is no point in keeping it up to date – apart from a slightly tidier profile.

So to summarize the git comments I propose a change like this: (assuming fetch+merge –ff-only)

To base you changes on the latest version of the [rclone master](https://github.com/rclone/rclone/tree/master) (upstream):

    git checkout master
    git fetch upstream
    git merge --ff-only
    git checkout my-new-feature
    git rebase master

I would then move this section just below “You may like one of the popular editors/IDE's for Go”.

And finally, I suggest rephrasing “rebase to master with `git rebase master`” to “base you code on the latest version of the upstream master”. (It is in the "Make sure to bullets" app. 15 lines below)

Thanks for the review, I hope my answers make sense 😅
(Perhaps you would like to review a new consolidated proposal?)

File: CONTRIBUTING.md (hunk 1 above)
Comment (ncw) [reply CONTRIBUTING.md:45]: That looks nice.

Note that on unix you'll have to run rclone in the current directory with

    ./rclone version

I'm not sure that syntax works on windows though

File: CONTRIBUTING.md (hunk 2 above)
Comment (ncw) [reply CONTRIBUTING.md:57]: I think your proposal is a good one. When I cast my mind back to when I was a git beginner it was having to deal with merge conflicts that caused me to do stuff like delete the whole repo and start again ;-)

> I’m also worried that (thoughtless) rebasing could lead to issues in later work. Suppose a beginner overlooks a misplaced commit in a successfully rebased master. Couldn’t that lead to troubles sometime later?

Rebasing either works, or gives a conflict. You can think of a rebase as saving your existing work as patches then reapplying them to the branch you are rebasing on. If there is a conflict in those patches you'll have to fix it.

So rebasing  should be harmless and any branches will need to be rebased eventually since rclone uses a [linear history model with no merges](https://www.bitsnbites.eu/a-tidy-linear-git-history/). The GithHub merge PR interface will do the rebase for me if there are no conflicts.

File: CONTRIBUTING.md (hunk 4 above)
Comment (ncw) [reply CONTRIBUTING.md:59]: Yes an update to the PR would be great.

You can do this in two ways.

1. Make the changes. Add in a new commit to the branch. Push the branch
2. Make the changes. Amend the existing commit (`git commit --amend`). Push with `--force` the branch.

Either will do. for 1. GithHub gives a consolidated view. 2. is the neatest way of doing it since you are presenting the actual commit for merging, wheras with 1. I'd squash the commit (also available in the GitHub UI) before merging.

File: CONTRIBUTING.md (hunk 1 above)
Comment (olefrost) [reply CONTRIBUTING.md:45]: I actually use the exact same syntax in PowerShell, so I will change. I don’t think there are many developers still using the good old CMD.

File: CONTRIBUTING.md (hunk 2 above)
Comment (olefrost) [reply CONTRIBUTING.md:57]: Thanks, and nice to learn more about rebasing, linear history and GitHub.

I still haven’t tried a merge conflict and do my very best to avoid it. I should probably try it out one day on a test branch to get more comfortable.

File: CONTRIBUTING.md (hunk 4 above)
Comment (olefrost) [reply CONTRIBUTING.md:59]: Sure, I will force push a squashed commit shortly; being both neat and curious to see how it affects the view/history here on GitHub.

I have changed as proposed in the above comments, expanded on using git and done some minor additions/changes to make the overall text/flow more fluent.

I hope you like the changes/additions, they will be pushed shortly.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -12,85 +12,132 @@ When filing an issue, please include the following information if
...
+    git push origin --follow-tags    # optional update of your fork in GitHub
+    git checkout my-new-feature
+    git rebase master
+
+### Squashing your commits ###
+
+To combine your commits into one commit:
+
+    git log                          # To count the commits to squash, e.g. the last 2
+    git reset --soft HEAD~2          # To undo the 2 latest commits
```
Comment (ncw) [inline CONTRIBUTING.md:129]: This isn't how exerienced git users do squashing, they'd use `git rebase -i master` which gives you the flexibility to just squash some things, keep and reword commit messages and re-order commits.

It might be too much for these instructions though!

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -12,85 +12,132 @@ When filing an issue, please include the following information if
...
 
+    cd folder/with/changed/files
     go test -v
 
 Note that you may need to make a test remote, e.g. `TestSwift` for some
 of the unit tests.
 
-Note the top level Makefile targets
-
-  * make check
```
Comment (ncw) [inline CONTRIBUTING.md:57]: I think (not 100% sure) that `make check` got lost from the instructions. We should probably change all uses of `make test` to `make quicktest` which is actually what the CI runs.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -12,85 +12,132 @@ When filing an issue, please include the following information if
...
+
+If everything is fine, then make the new combined commit:
+
+    git commit                       # To commit the undone commits as one
+
+otherwise, you may roll back using:
+
+    git reflog                       # To check that HEAD{1} is your previous state
+    git reset --soft 'HEAD@{1}'      # To roll back to your previous state
+### GitHub Continous Integration ###
```
Comment (ncw) [inline CONTRIBUTING.md:140]: Continuous is spelled wrong here (and in the ref at the bottom!)

**Review by ncw — COMMENTED** [review]
Great stuff! More comments on the new additions :-)

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -12,85 +12,132 @@ When filing an issue, please include the following information if
...
 Make sure you
 
+  * Add [unit tests](#testing) for a new feature.
   * Add [documentation](#writing-documentation) for a new feature.
-  * Follow the [commit message guidelines](#commit-messages).
-  * Add [unit tests](#testing) for a new feature
-  * squash commits down to one per feature
-  * rebase to master with `git rebase master`
+  * [Commit](#committing-your-work) using the [message guideline](#commit-messages).
+  * [Base your code on the latest master](#basing-your-work-on-the-latest-master).
```
[the remark this answers is not in the document]
Comment (olefrost) [reply CONTRIBUTING.md:77]: No, I just moved them to the top of the list :-)

File: CONTRIBUTING.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (olefrost) [reply CONTRIBUTING.md:140]: Added, just to be sure.

Comment (olefrost) [reply CONTRIBUTING.md:140]: Thanks, I have added [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker), but still miss some of the safety net form Word.

Fixed a few additional typos :-)

File: CONTRIBUTING.md (hunk 5 above)
Comment (olefrost) [reply CONTRIBUTING.md:129]: Great tip, interactive rebase is really powerful (if you know what you are doing ;-) This is the nice interface I get in VS Code with GitLens: https://www.youtube.com/watch?v=P5p71fguFNI

But yes, it is too much for a beginner. I also like the current guide because it introduces some the basic principles of Git and demonstrates the safety net in case of mistakes (similar to fetch+merge). Credit goes to jtagcat, I just reformatted and expanded a bit. On a side note: “git blame” really ought to be renamed to “git fame” 😊

I have added: Tip: You may like to use “git -i rebase master” if you are experienced or have a more complex situation

File: CONTRIBUTING.md (hunk 6 above)
Comment (olefrost) [reply CONTRIBUTING.md:57]: Well spotted, sorry.

I have tried to clarify the different levels of testing with subheadings and light rephrasing. I hope I got it right. I have only tried the quick go test  – the rest is based on your comments and reading the Makefile.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -12,85 +12,132 @@ When filing an issue, please include the following information if
...
+    git checkout master
+    git fetch upstream
+    git merge --ff-only
+    git push origin --follow-tags    # optional update of your fork in GitHub
+    git checkout my-new-feature
+    git rebase master
+
+### Squashing your commits ###
+
+To combine your commits into one commit:
```
[the remark this answers is not in the document]
Comment (olefrost) [reply CONTRIBUTING.md:126]: Good to know, then I propose we reduce complexity for the simple contributor (like me) by removing squash and rebase from “Make sure to” – and keep the instructions as something the reviewer (you) can request at you own discretion when things get too time-consuming/complicated.

The instructions for commit and commit message already implicitly guides the contributor towards making a single commit (per feature). Some will do it by themselves; some will need a hint during review, and some will be handled by you. 

You will get a new commit shortly, this time I won’t amend/squash – just to see the difference. 
I will be happy to clean up using an interactive rebase, when we are ready to merge 😊

**Comment by ivandeex** [comment]
Cross-post.
We have PR #5303 by @olefrost that improves the contributor documentation w.r.t. standard tools like git.
.We PR #4771 which strives for easing the setup of web-based dev environment. It lacks any docs however making it useless. It would be nice @olefrost and @k84shailendra could coordinate their efforts.

**Comment by ivandeex** [comment]
_Cross-post._

We have PR #5303 by @olefrost that improves the contributor documentation w.r.t. standard tools like git.

We also have PR #4771 which strives for easing the setup of web-based dev environment. However it lacks any docs making it rather useless.

It would be nice @olefrost and @k84shailendra could coordinate their efforts. Just an idea.

**Comment by olefrost** [comment]
I would like to keep the contributing guide independent of platform (Linux/Windows/MacOS,…), development environment (Local/VM/Docker/Kubernetes/GitPod/Codespaces/WSL/…), editor etc.
 
I also like that the guide focus on making a development environment that is completely free to establish and use (assuming you already have a computer with internet access).
 
The PR #4771 seems to promote a single commercial product and therefore violates the above principles. I have detailed my observations in [this comment to #4771 ](https://github.com/rclone/rclone/pull/4771#issuecomment-867780360).
 
What are your thoughts? (@ncw, @ivandeex, others)

**Comment by ncw** [comment]
> I would like to keep the contributing guide independent of platform (Linux/Windows/MacOS,…), development environment (Local/VM/Docker/Kubernetes/GitPod/Codespaces/WSL/…), editor etc.

That would be great, otherwise it will get too long and too difficult to keep up to date.
 
> I also like that the guide focus on making a development environment that is completely free to establish and use (assuming you already have a computer with internet access).
> 
> The PR #4771 seems to promote a single commercial product and therefore violates the above principles. I have detailed my observations in [this comment to #4771 ](https://github.com/rclone/rclone/pull/4771#issuecomment-867780360).

Agreed.

I'm going to merge this now - if there are any followups they can go in a new PR.

Apologies for the very long incubation period @olefrost - I'll try to do better next time!

**Comment by ncw** [comment]
Thank you for doing this @olefrost - it is much appreciated :-)

**Comment by olefrost** [comment]
@ncw Thank you - it’s been a joyful learning experience! :-)
